### ISSUE TITLE
Multiple strikes in OpenType bitmap fonts can't use different advance widths

### ISSUE THREAD

**Issue by lampmerchant** [harfbuzz/harfbuzz#4430]
Hi HarfBuzz devs!

tl;dr: EBDT/EBLC tables contain advance widths per strike in pixels, but HarfBuzz ignores them and only uses advance widths from the hmtx table

What I'm trying to do is create an OpenType (.otb) version of a bitmap font that contains multiple strikes.  The problem I have is that even in bitmap-only OpenType fonts, HarfBuzz appears to take advance widths from the hmtx table and ignore the ones provided in the EBDT/EBLC tables.  This wouldn't be a problem in a bitmap font with only one strike, since the values from EBDT/EBLC could just be mirrored in hmtx (after being converted from pixels to font units), however, in fonts with multiple strikes, if the advance widths aren't proportional to the pixels per em, the font writer appears to be out of luck...

I'm attaching a font with such a problem: [monegasque.otb.gz](https://github.com/harfbuzz/harfbuzz/files/12838159/monegasque.otb.gz)

It contains strikes at both 12 ppem and 16 ppem (9 and 12 point), however only the advance widths  of the 12 ppem strike are correct - the advance widths of the 16 ppem strike are too wide.

The 16 ppem strike looks like this:
![image](https://github.com/harfbuzz/harfbuzz/assets/66847847/e1652d25-6f24-46f5-8fec-e78ba25488b5)

But it should look like this:
![image](https://github.com/harfbuzz/harfbuzz/assets/66847847/47a32a03-9695-45f4-9a19-d498c36f2999)

Am happy to provide any more information that might be needed.  Thanks for looking!

**Comment by behdad** [comment]
We can implement that (and `hdmx` as well?) for when ppem is set.

**Comment by lampmerchant** [comment]
That'd be super cool. Thank you!

**Comment by behdad** [comment]
I don't have time immediately work on this. If anyone else wants to pick it up that would be great.

**Comment by lampmerchant** [comment]
Happy new year!  Just wanted to give this a bump - please let me know if there's anything else I can provide by way of debugging info to facilitate this.

**Comment by behdad** [comment]
I'll take a look soon.

**Comment by behdad** [comment]
One issue I can think of is that the `hdmx` numbers are designed for the font when hinting is enabled. That's problematic for renderers that don't do hinting. I suppose those then should not set `ppem`. But how to know to set `ppem` for bitmap fonts and not other fonts is not clear.

**Comment by lampmerchant** [comment]
Hi @behdad - just wanted to shake the tree on this a little.  I'm not sure how to answer the question about the `hdmx` numbers, only to say that my converter does not create an `hdmx` table (maybe it should?)  In any case, please let me know if I can provide any more information about the non-working font.

I know this probably isn't high on the priority list and I appreciate you taking a look. =)  Thanks!

**Comment by jafd** [comment]
Ah, so it may also be the cause of some bitmap fonts getting `s p a c e d  r e a l l y  w i d e`, especially when a higher-dpi application is trying to render them — font units are not necessarily the same as physical pixels. (I tried this with a single-strike font generated with the default FontForge settings; I'm not sure which tables are present there, will look when I have some time to play with it again.)
